### PR TITLE
Feature/principle of least surprise

### PR DESCRIPTION
**Pull request by samtrahan**
Makes `-a`, `-c`, `-t`, and `-m` options consistent across all Rocoto commands.  This required refactoring the option parsing and task/cycle selection code.  Manpages added for rocotorewind and rocotocomplete.  Other manpages are updated to reflect new option parsing.

### CODE REVIEW

**Comment by christopherwharrop** [comment]
This is a huge amount of changes.  It will take me a while to go thru this and evaluate it.

**Comment by samtrahan** [comment]
Let me know if you want to do something over the phone or Hangouts.  It may go faster if I explain things.

**Comment by christopherwharrop** [comment]
Since this is not related to the Slurm updates, let's leave it until after the holidays when I'm not being rushed to get so many things done.  After the other PRs are merged, you can merge develop onto this branch and continue to test and use it with all the other updates included.  I may give you a call next month to help get through it.

**Comment by samtrahan** [comment]
This branch is already tested on top of the other PRs.  It was tested on top of them in the slurm-of-least-surprise branch.  That branch should match the final "develop" other than the one "if" statement about `<memory>`.

**Comment by christopherwharrop** [comment]
Ok.  But, due to the huge number of changes involved, I want to spend some time looking at them before this gets merged.  I don't have that kind of time until next month.

**Comment by samtrahan** [comment]
That's fine.  I'll update feature/principle-of-least-surprise and have people keep testing it.  It is the version used by NEMS and the FV3 GFS workflows, so there will be plenty of testing.

**Comment by christopherwharrop** [comment]
Great!  I'm happy to hear about the testing.  Even though the test is showing positive results, I still need to understand what's going on with the code before merging it.  Thank you for your patience.

**Review by aerorahul — COMMENTED** [review]
Dear Developers,

Sorry to intrude on this pull request.
I am trying to use Rocoto on a NASA cluster Discover with Slurm scheduler, and I could not get develop to work, so I used this branch and it works except for the following syntax error.

I had to change line 282 in slurmbatchsystem.rb to:
queued_jobs = queued_jobs.encode('UTF-8', 'binary', {:invalid => :replace, :undef => :replace, :        replace => ''})

This was basically what is in develop, without the conditional if encode check.

Thanks for your attention.

**Comment by christopherwharrop** [comment]
@aerorahul, thanks for the note about that.  There has been some confusion about that line of code and the issue has to do with a Ruby 1.8.7 vs 2.x version compatibility conflict.  I thought I had it sorted out, but I'll take another look at it.  Which version of Ruby are you using on Discover?  The old "if" statement won't work because the offending code won't even parse in 1.8.7, so it's not just a question of checking the version and doing one thing or the other.

**Comment by aerorahul** [comment]
@christopherwharrop 

`1003 discover35 ~ $ ruby --version`
`ruby 1.8.7 (2011-12-28 patchlevel 357) [x86_64-linux]`

Unfortunately, discover is so far behind in its operating system state, that I had to comment out that line completely in order to get it to work. Otherwise, I get `STATUS` as `UNAVAILABLE` regardless of what `scontrol show job` says.

I have to add, I am no expert in `SLURM`, `ruby` OR `rocoto`, and am simply reacting to issues on Discover.

After that fix, I was able to submit, monitor, rewind and resubmit jobs on Discover.

Please let me know if I could be of any help, since I do wish to have Rocoto usable on Discover.

Thanks.

**Comment by christopherwharrop** [comment]
@aerorahul, Ok.  I suspected you were using 1.8.7.  We need to figure out what is going on.  I do have access to Ruby 1.8.7 on Jet and Theia (I can install it myself if needed) and so should be able to debug there.  But, as a precaution, I will probably ask you to try versions on Discover.  The reason for that line, and why I don't want to just get rid of it, is because it is necessary in order to correctly parse SLURM output on machines in other countries where there are unexpected characters.

You are correct, though, that this branch has the wrong version of that line and that it needs, at a minimum, to be corrected to the version that is now in develop.  As it is, this branch won't even parse on Ruby 1.8.7.

**Review by christopherwharrop — CHANGES_REQUESTED** [review]
Please merge develop onto this branch.  There are some things, like the encoding fix in slurmbatchsystem that are not here that should be here.

File: lib/wfmstat/statusengine.rb [hunk 1]
```diff
@@ -58,6 +58,7 @@ def initialize(options)
         # Get command line options
         @options=options
 
+
```
Comment (christopherwharrop) [inline lib/wfmstat/statusengine.rb]: Please remove extra blank lines that do not add clarity.

File: lib/wfmstat/statusengine.rb [hunk 2]
```diff
@@ -354,19 +321,24 @@ def print_status
 
         # Sort the task list in sequence order
         tasklist=jobs.keys | definedTasks.values.collect { |t| t.attributes[:name] }
-        unless @options.tasks.nil?
```
Comment (christopherwharrop) [inline lib/wfmstat/statusengine.rb:357]: Please delete the commented out lines and the extra blank line.

File: lib/wfmstat/statusengine.rb [hunk 3]
```diff
@@ -403,16 +375,26 @@ def print_status
...
+          if ! @subset.is_selected? cycle
+            #puts "#{cycle.class.name} #{cycle.inspect}: not selected"
+            next
+          end
+
           printf "================================================================================================================================\n"
 
           # Sort the task list in sequence order 
           tasklist=jobs.keys | definedTasks.values.collect { |t| t.attributes[:name] }
-          unless @options.tasks.nil?
```
Comment (christopherwharrop) [inline lib/wfmstat/statusengine.rb:410]: Please delete commented out lines.

File: lib/wfmstat/statusengine.rb [hunk 4]
```diff
@@ -403,16 +375,26 @@ def print_status
...
-            tasklist = tasklist.find_all { |task| @options.tasks.any? { |pattern| task=~/#{pattern}/ } }
-          end
+          #unless @options.tasks.nil?
+          #  tasklist = tasklist.find_all { |task| @options.tasks.any? { |pattern| task=~/#{pattern}/ } }
+          #end
           tasklist=tasklist.sort_by { |t| [definedTasks[t].nil? ? 999999999 : definedTasks[t].seq, t.split(/(\d+)/).map { |i| i=~/\d+/ ? i.to_i : i }].flatten }
           tasklist.each do |task|
 
+            if ! @subset.is_selected? task
+              #puts "#{task}: not selected"
```
Comment (christopherwharrop) [inline lib/wfmstat/statusengine.rb:386]: Please delete commented out line

File: lib/workflowmgr/cycle.rb [hunk 5]
```diff
@@ -57,18 +57,29 @@ def initialize(cycle,params={ :activated=>Time.at(0), :expired=>Time.at(0), :don
...
+    ##########################################
+    #
+    # hash
+    #
+    ##########################################
+    def hash()
+      return @cycle.hash
+    end
+
+
```
Comment (christopherwharrop) [inline lib/workflowmgr/cycle.rb:82]: One too many blank lines.

File: lib/wfmstat/wfmstatoption.rb [hunk 6]
```diff
@@ -5,147 +5,60 @@
 ##########################################
 module WFMStat
 
+  require 'workflowmgr/workflowsubsetoptions'
+  require 'workflowmgr/workflowdbselection'
+
   ##########################################  
   #
   # Class StatusOption
```
Comment (christopherwharrop) [inline lib/wfmstat/wfmstatoption.rb:13]: Change comment to match class name

File: lib/wfmstat/wfmstatoption.rb [hunk 7]
```diff
@@ -5,147 +5,60 @@
...
 
-     end  # parse
+    end # add_opts
+
+    def make_selection()
+      @selection=WorkflowMgr::WorkflowDBSelection.new(@all_tasks,@task_options,@cycles,@default_all)
+      return @selection
+    end
 
   end  # Class StatusOption
```
Comment (christopherwharrop) [inline lib/wfmstat/wfmstatoption.rb:63]: Change comment to match class name

File: lib/workflowmgr/cycle.rb [hunk 8]
```diff
@@ -80,6 +91,18 @@ def <=>(other)
...
+    ##########################################
+    #
+    # inactive?
+    #
+    ##########################################
+    def inactive?
+
+      return @state==:inactive
+
+    end  # active?
```
Comment (christopherwharrop) [inline lib/workflowmgr/cycle.rb:102]: Change comment to match method name

File: lib/workflowmgr/cycle.rb [hunk 9]
```diff
@@ -157,11 +180,28 @@ def reactivate!
...
+    ##########################################
+    def rewind!
+
+      @done=Time.at(0)
+      @draining=Time.at(0)
+      @expired=Time.at(0)
+      @activated=Time.at(0)
+      @state=:inactive
+
+    end  # activate!
```
Comment (christopherwharrop) [inline lib/workflowmgr/cycle.rb:201]: Change comment to match method name

File: lib/workflowmgr/selectionutil.rb [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  #
+  ##########################################
+  class MetataskSelection
+    # Stores the contents of a -m option from argument parsing
+    attr_reader :arg
+    def initialize(arg) @arg=arg ; end
+  end # class MetataskSelection
+
+
+
```
Comment (christopherwharrop) [inline lib/workflowmgr/selectionutil.rb]: One too many blank lines

File: lib/workflowmgr/workflowdoc.rb [hunk 11]
```diff
@@ -318,10 +318,8 @@ def cycledefs
       }
 
       return cycles
-
```
Comment (christopherwharrop) [inline lib/workflowmgr/workflowdoc.rb:321]: Please don't remove blank lines unless you have a reason.

File: lib/workflowmgr/workflowengine.rb [hunk 12]
```diff
@@ -91,12 +91,9 @@ def initialize(options)
     #
     ##########################################
     def rewind!
+
```
Comment (christopherwharrop) [inline lib/workflowmgr/workflowengine.rb:94]: One too many blank lines

File: lib/workflowmgr/workflowengine.rb [hunk 13]
```diff
@@ -116,59 +113,77 @@ def rewind!
...
+            #puts "Rewind specified tasks."
             task_list=@options.tasks
           end
 
           rewind_list=[]
           did_something=false
-          puts "ACTIVE JOBS: <#{@active_jobs.keys.join('><')}>"
-          task_list.each do |task_name|
-            puts "#{strcyc}: #{task_name}: consider this task"
+          #puts "ACTIVE JOBS: <#{@active_jobs.keys.join('><')}>"
```
Comment (christopherwharrop) [inline lib/workflowmgr/workflowengine.rb:175]: Remove commented line

File: lib/workflowmgr/workflowengine.rb [hunk 14]
```diff
@@ -116,59 +113,77 @@ def rewind!
...
+          rewind_tasks.each do |task_name|
+            #puts "#{strcyc}: #{task_name}: consider this task"
```
Comment (christopherwharrop) [inline lib/workflowmgr/workflowengine.rb:177]: Remove commented line

File: lib/workflowmgr/workflowengine.rb [hunk 15]
```diff
@@ -116,59 +113,77 @@ def rewind!
...
             cycjob=@active_jobs[task_name.to_s]
             if cycjob.nil?
-              puts "#{strcyc}: #{task_name}: job has not been tried yet.  Doing nothing to this task."
+              #puts "#{strcyc}: #{task_name}: job has not been tried yet.  Doing nothing to this task."
```
Comment (christopherwharrop) [inline lib/workflowmgr/workflowengine.rb:180]: Remove commented line

File: lib/workflowmgr/workflowengine.rb [hunk 16]
```diff
@@ -116,59 +113,77 @@ def rewind!
...
               next
             end
-            puts "#{strcyc}: #{task_name}: jobs exist for <#{cycjob.keys.join(', ')}>"
+            #puts "#{strcyc}: #{task_name}: jobs exist for <#{cycjob.keys.join(', ')}>"
```
Comment (christopherwharrop) [inline lib/workflowmgr/workflowengine.rb:183]: Remove commented line

File: lib/workflowmgr/workflowengine.rb [hunk 17]
```diff
@@ -116,59 +113,77 @@ def rewind!
...
             rewind_job=cycjob[cycle]
             if rewind_job.nil?
-              puts "#{strcyc}: #{task_name}: job has not been tried yet for cycle #{strcyc}.  Doing nothing to this task."
+              #puts "#{strcyc}: #{task_name}: job has not been tried yet for cycle #{strcyc}.  Doing nothing to this task."
```
Comment (christopherwharrop) [inline lib/workflowmgr/workflowengine.rb:186]: Remove commented line

File: lib/workflowmgr/workflowengine.rb [hunk 18]
```diff
@@ -199,9 +214,27 @@ def rewind!
...
+
+          if do_not_reactivate
+            puts "#{strcyc}: not reactivating this cycle."
+          elsif all_tasks
+            rewind_cycle.rewind!
+
+
+            puts "#{strcyc}: Deactivate cycle: #{rewind_cycle}"
+
+            # @dbServer.update_cycles([rewind_cycle])
```
Comment (christopherwharrop) [inline lib/workflowmgr/workflowengine.rb]: Remove commented line

File: lib/workflowmgr/workflowengine.rb [hunk 19]
```diff
@@ -695,17 +712,18 @@ def force_complete!
...
           } # complete_tasks.each
+
         } # complete_cycles.each
-      } # with_locked_db
 
-              # Deactivate completed cycles
+        # Deactivate completed cycles
         deactivate_done_cycles
-
+
```
Comment (christopherwharrop) [inline lib/workflowmgr/workflowengine.rb]: Blank lines should not have tabs in them

File: lib/workflowmgr/workflowselection.rb [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+        end
+      end
+      tasks=selection.to_a
+      tasks.sort!
+      return tasks
+    end
+
+    ##########################################  
+    #
+    # select_tasks
```
Comment (christopherwharrop) [inline lib/workflowmgr/workflowselection.rb]: Change comment to match method name

File: lib/workflowmgr/workflowsubsetoptions.rb [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+    #
+    ##########################################
+    def add_task_option(opt)
+      @task_options << opt
+    end
+
+
+    ##########################################
+    #
+    # validate_args
```
Comment (christopherwharrop) [inline lib/workflowmgr/workflowsubsetoptions.rb]: Change comment to match method name

File: sbin/rocotoboot.rb [hunk 22]
```diff
@@ -15,14 +15,15 @@
...
+require 'workflowmgr/workflowsubsetoptions'
 require 'libxml'
 
 # Turn off that ridiculous Libxml-ruby handler that automatically sends output to stderr
 # We want to control what output goes where and when                                     
 LibXML::XML::Error.set_handler(&LibXML::XML::Error::QUIET_HANDLER)
 
 # Create workflow status and run it
-workflowEngine=WorkflowMgr::WorkflowEngine.new(WorkflowMgr::WorkflowBootOption.new(ARGV))
+opt=WorkflowMgr::WorkflowSubsetOptions.new(ARGV,'rocotorewind','rewind')
```
Comment (christopherwharrop) [inline sbin/rocotoboot.rb]: Replace "rewind" with "boot"

**Review by christopherwharrop — CHANGES_REQUESTED** [review]

**Comment by samtrahan** [comment]
All trailing whitespace in Rocoto Ruby code has been removed and indentation tabs are now spaces.  I've removed commented-out code from many places.  The "develop" branch has been merged into this feature branch.  So far as I know, this feature branch is up-to-date with all bug fixes and documentation updates to the feature.

**Review by christopherwharrop — APPROVED** [review]
Thank you for fixing all the whitespace issues.

**Pull request merged**